plugins/auth/irods_auth_plugin-native/libirods_auth_plugin-native.cpp

```diff
@@ -0,0 +1,467 @@
+            return;
+        }
+
+        // This is a proxy user attempting to authenticate a client user with insufficient
+        // permissions to do so.
+        THROW(SYS_PROXYUSER_NO_PRIV, fmt::format(
+              "proxy user: [{}] with {} no priv to auth client user: [{}].",
+              _comm->proxyUser.userName, _proxy_priv, _comm->clientUser.userName));
+    } // throw_on_insufficient_privilege_for_proxy_user
+#endif
+} // anonymous namespace
+
+namespace irods
+{
+    class native_authentication : public irods::experimental::authentication_base {
+    public:
+        native_authentication()
+        {
+            add_operation(AUTH_ESTABLISH_CONTEXT,    OPERATION(rcComm_t, native_auth_establish_context));
+            add_operation(AUTH_CLIENT_AUTH_REQUEST,  OPERATION(rcComm_t, native_auth_client_request));
+            add_operation(AUTH_CLIENT_AUTH_RESPONSE, OPERATION(rcComm_t, native_auth_client_response));
+            #ifdef RODS_SERVER
+            add_operation(AUTH_AGENT_START,          OPERATION(rsComm_t, native_auth_agent_start));
+            add_operation(AUTH_AGENT_AUTH_REQUEST,   OPERATION(rsComm_t, native_auth_agent_request));
+            add_operation(AUTH_AGENT_AUTH_RESPONSE,  OPERATION(rsComm_t, native_auth_agent_response));
+            add_operation(AUTH_AGENT_AUTH_VERIFY,    OPERATION(rsComm_t, native_auth_agent_verify));
+            #endif
+        } // ctor
+
+    private:
+        json auth_client_start(rcComm_t* comm, const json& req)
+        {
+            json resp{req};
+            resp["next_operation"] = AUTH_CLIENT_AUTH_REQUEST;
+            resp["user_name"] = comm->proxyUser.userName;
+            resp["zone_name"] = comm->proxyUser.rodsZone;
+
+            return resp;
+        } // auth_client_start
+
+        json native_auth_establish_context(rcComm_t* comm, const json& req)
+        {
+            json resp{req};
+
+            if(auto [e, n] = verify(req, {"user_name","zone_name","request_result"}); !e) {
+                THROW(SYS_INVALID_INPUT_PARAM,
+                      fmt::format("[{}:{}] missing [{}] in request", __func__, __LINE__, n));
+            }
+
+            auto request_result = req.at("request_result").get<std::string>();
+            request_result.resize(CHALLENGE_LEN);
+
+            // build a buffer for the challenge hash
+            // Buffer structure:
+            //  [64+1] challenge string (generated by server and sent back here)
+            //  [50+1] obfuscated user password
+            char md5_buf[CHALLENGE_LEN + MAX_PASSWORD_LEN + 2]{};
+            strncpy(md5_buf, request_result.c_str(), CHALLENGE_LEN);
+
+            // Save a representation of some of the challenge string for use as a session signature
+            setSessionSignatureClientside(md5_buf);
+
+            // determine if a password challenge is needed, are we anonymous or not?
+            bool need_password = false;
+            if (req.at("user_name") == ANONYMOUS_USER) {
```

Comment:
let's use an explicit cast as recommended by the nlohmann::json library

Reply:
Would it do to use `req.at("user_name").get_ref<std::string&>()`?

Reply:
that is correct. using the `get<T>()` member functions is the explicit form for the library.

you'll need to add a `const` to that template argument though.

Reply:
Cool, thanks!